File: src/rout.c
/*      PROGRAM rout, C-version   

     Routing algorithm developed by D. Lohmann.
     Rewritten from FORTRAN to C 2003, AWW/IH.

     New features, 2003 (IH):
      Upstream routed areas are not rerouted
      Dams/reservoirs included (based on power or streamflow demand)

     MAXROWS and MAXCOLS should be larger than the grid
     MAXYEARS should equal at least run length yrs+1   
     i: row from bottom (starts at 1)
     j: col from left (starts at 1)
     [][]: [row][col]
     Indexing (arrays): Starts at 1 (i.e. 0 is left unused)

make

*/

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <float.h>
#include <math.h>
#include <string.h>
#include "rout.h"
/*************************************************************/
/* Start of ROUT                                             */
/*************************************************************/
int main (int argc, char *argv[]) {   

  FILE *fp;

  ARC **BASIN=NULL; //Grid input information is stored here,
                    //like direction,fraction,velocity of 
                    //each cell in the grid. Row(i) is row from
                    //the bottom, col(j) is column from the left.
                    //Numbering starts at 1 for both i and j.
  LIST *STATION=NULL;  //Information about station locations and names.
                    //Area is included in the list, although the value
                    //isn't used in the routing program.
  TIME *DATE=NULL;

  char *filename;
  char *filename_reservoirs;
  char *inpath;
  char *fluxpath;
  char *demandpath;
  char *outpath;
  char *workpath;
  char *naturalpath;
  char *dummy;
  char *name;
  char uhstring[20][20]; //name of uhfile (or "NONE")

  float xllcorner; //x-coordinate, lower left corner of grid
  float yllcorner; //y-coordinate, lower left corner of grid
  float size;      //size of grid cell, in degrees
  float lat;
  float lon;
  float value;
  float test;
  float factor_sum;
  float ***UH;     /* Impulse response function UH[row][col][48] */
                   /* Based on Lohmann's Tellus article          */
		   /* Depends on velocity, diffusion and size    */  
  float **UH_BOX;  /* Unit hydrograph[number_of_cells][12]       */
                   /* Normally the 12 numbers are equal to       */
                   /* the ones in the unit hydrograph file       */ 
  float **UH_DAILY; /* UH_DAILY[number_of_cells][uh_day]         */
  float **UH_S;    /* .uh_s grid, UH_S[numberofcells][ke+uh_day] */
  float **FR;
  double *BASEFLOW;  
  double *RUNOFF;
  double *FLOW;
  float *R_FLOW;
  float *STORAGE;
  float *LEVEL;
  float *PowerProd; /* Numbers from moscem, unit MW */
  float **RESEVAPDATA;
  float **WATER_DEMAND; /* Water demand (irrigation) for reservoirs */

  int **CATCHMENT; /*A list with row and col number for all cells upstream 
                     station location (includes station location). 
                     CATCHMENT[cellnumber][row(0)/col(1)/routed(2)]
                     The order of the list is arbitrary. 
                     CATCHMENT[cellnumber][2]=0 means routing as normal
                     CATCHMENT[cellnumber][2]=1 means this cell is routed before, in current routing
                     CATCHMENT[cellnumber][2]=2 means flow already exist for that
                     cell (routed some time previously). */
  int i,j;
  int nr;
  int nrows,ncols;     //number of rows/columns in basin 
                       //(read from direction file)
  int active_cells;    //total number of active cells in grid
  int number_of_cells; //number of cells upstream a station location
  int upstream_cells;			   
  int number_of_stations; //number of stations 
  int decimal_places;  //decimal places in VIC output 
  int start_year,start_month; //start of VIC simulation
  int stop_year,stop_month;
  int first_year,first_month; //start of output to be written
  int last_year,last_month;
  int ndays,nmonths;   //number of days and months to be routed 
  int skip;            //number of days to skip in VIC simulations
  int irr_rout;        //irrigation type routing    
  int res_rout;        //reservoir routing should be included	   
  int irow,icol;			   
  int missing;
  int demand;
  int basin_number;    //basin number...  
  int nbytes=54; //nbytes pr day in flux files used in ReadDataForReservoirEvaporation. bad programming....should not be hardcoded!!! Tian changed it to 54 Apr 2013
  /***********************************************************/

  if(argc != 2){
   printf("USAGE:  rout <infile>\n");
   exit(0);
  }
  if((fp = fopen(argv[1], "r")) == NULL) {
   printf("Cannot open %s\n",argv[1]);
   exit(1);
  }

  /* Allocate memory for input parameters/names */
  filename = (char*)calloc(100,sizeof(char));
  filename_reservoirs = (char*)calloc(100,sizeof(char));
  fluxpath = (char*)calloc(100,sizeof(char));
  inpath = (char*)calloc(100,sizeof(char));
  demandpath = (char*)calloc(150,sizeof(char));
  outpath = (char*)calloc(100,sizeof(char));
  workpath = (char*)calloc(100,sizeof(char));
  naturalpath = (char*)calloc(100,sizeof(char));
  dummy = (char*)calloc(100,sizeof(char));
  name = (char*)calloc(100,sizeof(char));

  /* Find basin number  */
  fgets(dummy, MAXSTRING, fp);
  fscanf(fp, "%*s %d",&basin_number);

  /* Find number of rows and cols in grid of interest */
  fscanf(fp, "%*s %s",filename);
  FindRowsCols(filename,&nrows,&ncols,&xllcorner,
               &yllcorner,&size,&missing);

  /* Allocate memory for BASIN, UH, FR */
  BASIN=calloc(nrows+1,sizeof(ARC));
  for(i=0;i<=nrows;i++) 
      BASIN[i]=calloc(ncols+1,sizeof(ARC));
  UH = (float***)calloc(nrows+1,sizeof(float**));
  for(i=0;i<=nrows;i++) {
    UH[i]=(float**)calloc(ncols+1,sizeof(float*));
    for(j=0;j<=ncols;j++) 
      UH[i][j]=(float*)calloc(LE+1,sizeof(float*));      
  }  
  FR = (float**)calloc(TMAX+1,sizeof(float*));
  for(i=0;i<=TMAX;i++) 
     FR[i]=(float*)calloc(2,sizeof(float));


  /* Read direction file */
  printf("Direction file: %s\n",filename);
  ReadDirection(filename,BASIN,nrows,ncols,&active_cells,missing);
  printf("Active cells in basin: %d\n",active_cells);

  /* Allocate memory for CATCHMENT, UH_BOX, UH_S, 
                         UH_DAILY, STATION */
  CATCHMENT = (int**)calloc(active_cells+1,sizeof(int*));
  for(i=0;i<=active_cells+1;i++) 
    CATCHMENT[i]=(int*)calloc(4,sizeof(int));
  UH_BOX=(float**)calloc(active_cells+1,sizeof(float*));
  for(i=0;i<=active_cells+1;i++) 
      UH_BOX[i]=(float*)calloc(KE+1,sizeof(float*));      
  UH_DAILY=(float**)calloc(active_cells+1,sizeof(float*));
  for(i=0;i<=active_cells+1;i++) 
      UH_DAILY[i]=(float*)calloc(UH_DAY+1,sizeof(float));      
  UH_S=(float**)calloc(active_cells+1,sizeof(float*));
  for(i=0;i<=active_cells+1;i++) 
      UH_S[i]=(float*)calloc(KE+UH_DAY+1,sizeof(float));      
  STATION=calloc(20,sizeof(LIST));
  DATE=calloc(DAYS,sizeof(TIME));

  /* Read velocity file if any */
  fscanf(fp, "%*s %s",filename);
  value=atof(filename);
  if(value<EPS) ReadVelocity(filename,BASIN,nrows,ncols);
  else {
    printf("Velocity: %.2f\n",value);
    for(i=1;i<=nrows;i++) 
      for(j=1;j<=ncols;j++) 
	BASIN[i][j].velocity=value;
  }

  /* Read diffusion file */
  fscanf(fp, "%*s %s",filename);
  value=atof(filename);
  if(value<EPS) ReadDiffusion(filename,BASIN,nrows,ncols);
  else {
    printf("Diffusion: %.2f\n",value);
    for(i=1;i<=nrows;i++)
      for(j=1;j<=ncols;j++) 
        BASIN[i][j].diffusion=value;
  }

  /* Read xmask file */
  fscanf(fp, "%*s %s",filename);
  value=atof(filename);
  //if(value<EPS) 
  ReadXmask(filename,BASIN,nrows,ncols);
  /*else {
    printf("Xmask: %.2f\n",value);
    for(i=1;i<=nrows;i++)
      for(j=1;j<=ncols;j++) 
        BASIN[i][j].xmask=value;
	}*/

  /* Read fraction file */
  fscanf(fp, "%*s %s ",filename);
  value= atof(filename);
  //if(value<EPS) 
  ReadFraction(filename,BASIN,nrows,ncols);
  /*else {
    printf("Fraction: %.2f filename: %s %f\n",value,filename);
    for(i=1;i<=nrows;i++)
      for(j=1;j<=ncols;j++) 
        BASIN[i][j].fraction=value;
	}*/

  /* Read routing information 
     irr_rout=0: Nonirrigated part of cell, or entire cell
     irr_rout=1: Irrigated part of cell */
  fscanf(fp, "%*s %d",&irr_rout); 

  /* Read information on previously routed grid cells */
  fscanf(fp, "%*s %s",filename);
  value=atof(filename);
  if(value<EPS) {
    ReadRouted(filename,BASIN,nrows,ncols);
  }
  else {
    for(i=1;i<=nrows;i++)
      for(j=1;j<=ncols;j++) 
        BASIN[i][j].routed=0;
    printf("Routed cells: %s not found, setting values to zero\n",
	   filename);    
  }

  /* Include reservoir routing? 1: yes, 0: no */ 
  fscanf(fp, "%*s %d",&res_rout); 

  /* Read reservoir file name */
  fscanf(fp, "%*s %s",filename_reservoirs);
  printf("Reservoir file: %s\n",filename_reservoirs);
  value=atof(filename_reservoirs);
  if(value<EPS && res_rout==1) ReadReservoirs(filename_reservoirs,nrows,ncols,BASIN);
  else {
    printf("Reservoirs not taken into account\n");
    for(i=1;i<=nrows;i++)
      for(j=1;j<=ncols;j++) 
        BASIN[i][j].resloc=0;
  }

  /* Read station file */
  fscanf(fp, "%*s %s",filename);
  printf("Station file: %s\n",filename);
  ReadStation(filename,BASIN,STATION,nrows,ncols,
	      uhstring,&number_of_stations);
  printf("Station file: %s %d\n",filename,number_of_stations);
  /* Read input paths and precision of VIC filenames, and output path */
  fscanf(fp, "%*s %s",fluxpath); 
  fscanf(fp, "%*s %s",inpath);
  fscanf(fp, "%*s %d",&decimal_places);
  fscanf(fp, "%*s %s",demandpath); //filepath to water demands, reservoirs
  fscanf(fp, "%*s %d",&demand);   //0/1: do not/do take demands into account
  fscanf(fp, "%*s %s",outpath);
  fscanf(fp, "%*s %s",workpath);
  fscanf(fp, "%*s %s",naturalpath); //filepath to routing results, naturalized simulations

  printf("Input file path: %s\n",inpath);
  printf("Decimal places: %d\n",decimal_places);
  printf("Demand path: %s\n",demandpath);
  printf("Demand: %d\n",demand);
  printf("Output file path: %s\n",outpath);
  printf("Working files path: %s\n",workpath);
  printf("Naturalized simulations: %s\n",naturalpath);

  /* Read start and end year/month from VIC simulation */
  fscanf(fp, "%*s %d %d %d %d",
	 &start_year,&start_month,&stop_year,&stop_month);

  /* Read start and end year/month for writing output */
  fscanf(fp, "%*s %d %d %d %d",
	 &first_year,&first_month,&last_year,&last_month);
  printf("Start input: %d %d  Start output: %d %d  \n",
	 start_year,start_month,first_year,first_month);  

  /* Calculate number of days & months to be routed, 
     and number of days to skip when reading VIC simulation 
     results */
  CalculateNumberDaysMonths(start_year,start_month,
                            first_year,first_month,
                            last_year,last_month, 
                            &skip,&ndays,&nmonths);
  printf("Output Start: %d %d  End: %d %d  Skip: %d Ndays: %d Nmonths:%d\n",
	 first_year,first_month,last_year,last_month,skip,ndays,nmonths);  

  /* Allocate memory for BASEFLOW, RUNOFF and OUTPUT arrays */
  RUNOFF=(double*)calloc(ndays+1,sizeof(double));
  BASEFLOW=(double*)calloc(ndays+1,sizeof(double));
  FLOW=(double*)calloc(ndays+1,sizeof(double));
  R_FLOW=(float*)calloc(ndays+1,sizeof(float));
  STORAGE=(float*)calloc(ndays+1,sizeof(float));
  LEVEL=(float*)calloc(ndays+1,sizeof(float));
  PowerProd=(float*)calloc(ndays+1,sizeof(float));
  RESEVAPDATA=(float**)calloc(ndays+1,sizeof(float*));
  for(i=0;i<ndays+1;i++) 
    RESEVAPDATA[i]=(float*)calloc(5,sizeof(float));   
  WATER_DEMAND=(float**)calloc(stop_year-start_year+1,sizeof(float*));
  for(i=0;i<stop_year-start_year+1;i++) 
    WATER_DEMAND[i]=(float*)calloc(13,sizeof(float));   

  /* Read name of uh-file */
  fscanf(fp, "%*s %s",filename);
  fclose(fp);

  /* Make impulse response function (UH). */
  /* Based on Lohmann's Tellus article    */
  printf("Making impulse response function.....UH[row][col][48]\n");
  MakeUH(UH,BASIN,nrows,ncols);

  /* Loop over required output stations, 
     rout fluxes and write to output files */  
  for(nr=1;nr<=number_of_stations;nr++) {
    for(j=1;j<=ndays;j++) {
	R_FLOW[j]=0.;
    }
    if(STATION[nr].id==1) {
        printf("\n\nSearching catchment..Location: row %d col %d\n",
               STATION[nr].row,STATION[nr].col);
        SearchCatchment(BASIN,CATCHMENT,STATION[nr].row,
                        STATION[nr].col,nrows,ncols,
                        STATION[nr].type,&number_of_cells,
			&upstream_cells);

        /* Read grid UH, UH_BOX[number_of_cells][12] */
        printf("Read grid UH_BOX...%s\n",filename);
        ReadGridUH(filename,UH_BOX,number_of_cells,
                   CATCHMENT);

        /* Make .uh_s-file if it doesn't exist, 
			     UH_S[numberofcells][ke+uh_day] */
        printf("Make grid UH_S...\n");
        MakeGridUH_S(BASIN,CATCHMENT,STATION,number_of_cells,nr,
	           UH_DAILY,UH,FR,UH_BOX,UH_S,uhstring[nr]);
 
        /* Make convolution. This is where the VIC fluxes are read. */
        printf("Make convolution...\n");
        MakeConvolution(number_of_cells,skip,ndays,CATCHMENT,
                        BASIN,BASEFLOW,RUNOFF,FLOW,
  		        UH_S,STATION,nr,
			xllcorner,yllcorner,size,
		        inpath,outpath,workpath,decimal_places,DATE,
			&factor_sum,first_year,first_month,irr_rout,
			missing);

	/* Find lat and lon of station */
	lat=yllcorner + STATION[nr].row*size - size/2.0;
        lon=xllcorner + STATION[nr].col*size - size/2.0;

	/* Reservoir routing if needed */
	if(!res_rout && !irr_rout) STATION[nr].type=1;
	if(STATION[nr].type==2) {
	  printf("Reservoir routing.....\n");
	  ReadDataForReservoirEvaporation(fluxpath,RESEVAPDATA,
	  				  lat,lon,skip,nbytes,ndays,
	  				  decimal_places); 
	  /* NB! Above function assumes daily time step!!! must be customized for your use!!! Or made flexible.... */
	  ReservoirRouting(filename_reservoirs,demandpath,
			   FLOW,R_FLOW,STORAGE,LEVEL,PowerProd,
			   WATER_DEMAND,RESEVAPDATA,DATE,STATION[nr].name,
	  		   STATION[nr].row,STATION[nr].col,
			   start_year,stop_year,ndays,demand,
			   basin_number,naturalpath);
	}

        /* Write data */
	printf("Writing data...\n");
        strcpy(name,STATION[nr].name);
        WriteData(FLOW,R_FLOW,STORAGE,LEVEL,PowerProd,
		  name,outpath,ndays,nmonths,
		  DATE,factor_sum,
                  first_year,first_month,last_year,
		  last_month,start_year,lat,lon,STATION[nr].type,
		  decimal_places);
	if(irr_rout==1) {
	  for(i=3;i<=upstream_cells;i++) {
	    irow=CATCHMENT[i][0];
	    icol=CATCHMENT[i][1];
	    BASIN[irow][icol].routed=1;	
	  }
	}
    }
  }

  /* Make new 'routed' file */
  MakeRoutedFile(BASIN,CATCHMENT,nrows,ncols,number_of_cells); 
 
  /* Make new direction file */
  MakeDirectionFile(BASIN,CATCHMENT,nrows,ncols,number_of_cells,
		    xllcorner,yllcorner,size,missing); 

  /* Free memory */
  for(i=0;i<=nrows;i++) {
    free(BASIN[i]);
    free(UH[i]);
   }
  free(BASIN);
  free(UH);

  /*  for(i=0;i<=active_cells+1;i++) 
    free(CATCHMENT[i]);
  free(CATCHMENT); 

  free(DATE);
  free(STATION);
  free(RUNOFF);
  free(BASEFLOW);
  free(FLOW);
  free(R_FLOW);
  free(STORAGE);
  free(LEVEL);*/

  return 0;
}

File: src/rout.f
      PROGRAM rout
c
c     Routing algorithm developed by D. Lohmann.
c
c     Modified to allow more flexible array dimensions and
c     the removal of harcoded file names.
c
c     Code maintained by G. O'Donnell (tempgd@hydro.washington.edu)
c     See WA Hydrology Homepage for operational details.

c     Modified 5/99 to read in the uh_s array if it has already
c     been generated in a previous run.

c     Modified 2/2001 by edm to include month and year in output
c     and also check dates in VIC output files and calculate NDAYS
c
      IMPLICIT NONE

      integer IARGC

      integer isaleap
      external isaleap

c     change dimensions here
c     nrow and ncol should be larger than the grid
c     nyr should equal run length yrs+1
      INTEGER NROW, NCOL, DAYS, NYR
      PARAMETER (NROW = 250, NCOL = 250)
      PARAMETER (NYR = 50)

c     no changes after here
      REAL    DT
      INTEGER KE, LE, TMAX, UH_DAY, PMAX
      PARAMETER (DAYS=NYR*366)
      PARAMETER (KE   = 12    )
      PARAMETER (LE   = 48    )
      PARAMETER (DT   = 3600.0)
      PARAMETER (UH_DAY = 96  )
      PARAMETER (TMAX = UH_DAY*24)
      PARAMETER (PMAX = 10000   )

      INTEGER DIREC(NCOL,NROW,2)
      REAL    VELO(NCOL,NROW), DIFF(NCOL,NROW)
      REAL    XMASK(NCOL,NROW), FRACTION(NCOL,NROW)
      REAL    UH_BOX(PMAX,KE), UHM(NCOL,NROW,LE)
      REAL    UH_S(PMAX,KE+UH_DAY-1)
      REAL    BASE(DAYS), RUNO(DAYS), FLOW(DAYS)

      INTEGER IDAY(DAYS), IMONTH(DAYS), IYEAR(DAYS)
      INTEGER MO(12*NYR),YR(12*NYR)
      INTEGER NO_OF_BOX
      INTEGER CATCHIJ(PMAX,2)
      INTEGER H(NCOL,NROW)

      INTEGER PI, PJ
      REAL    UH_DAILY(PMAX,UH_DAY)
      REAL    FR(TMAX,2)

      INTEGER NR
      INTEGER IROW, ICOL
      INTEGER LP,M,Y
      INTEGER J

      CHARACTER*80 UH_STRING         ! new, AW
      CHARACTER*21 NAME
      CHARACTER*5  NAME5
      CHARACTER*72 FILE_INPUT, FILENAME
      CHARACTER*72 INPATH, OUTPATH

      INTEGER DPREC

      REAL    AREA, FACTOR_SUM

      REAL XC, YC, SIZE

      REAL FDUM
      LOGICAL TORF
      INTEGER NDAY
      INTEGER NMONTHS

C**** variables for monthly means *****************************

      INTEGER DAYS_IN_MONTH(12)
      DATA DAYS_IN_MONTH /31,28,31,30,31,30,31,31,30,31,30,31/

      INTEGER START_YEAR, STOP_YEAR, FIRST_YEAR, LAST_YEAR
      INTEGER START_MO, STOP_MO, FIRST_MO, LAST_MO

      REAL MONTHLY(12*NYR)
      REAL MONTHLY_mm(12*NYR)
      REAL YEARLY(12)
      REAL YEARLY_mm(12)

C***********************************************************
C     OPEN NECESSARY FILES
C***********************************************************

c     INPUT FILE READ
c     process commandline args
      IF(IARGC().NE.1)THEN
         PRINT*, 'USAGE:  rout <infile>'
         STOP
      ENDIF
      CALL GETARG(1,FILE_INPUT)
      OPEN(1,FILE=FILE_INPUT,STATUS='OLD',ERR=9001)
      READ(1,'(//A)') FILENAME
      CALL READ_DIREC(DIREC,NCOL,NROW,H,XC,YC,SIZE,
     $     FILENAME,IROW,ICOL)
c     velo file
      READ(1,*)
      READ(1,*) TORF
      IF(TORF)THEN
         READ(1,'(A)') FILENAME
         CALL READ_VELO(VELO,NCOL,NROW,FILENAME,
     $        IROW,ICOL)
      ELSE
         READ(1,*) FDUM
         CALL INIT_ARRAY(VELO,NCOL,NROW,FDUM)
      ENDIF
c     diff file
      READ(1,*)
      READ(1,*)TORF
      IF(TORF)THEN
         READ(1,'(A)') FILENAME
         CALL READ_DIFF(DIFF,NCOL,NROW,FILENAME,
     $        IROW,ICOL)
      ELSE
         READ(1,*) FDUM
         CALL INIT_ARRAY(DIFF,NCOL,NROW,FDUM)
      ENDIF
c     xmask file
      READ(1,*)
      READ(1,*)TORF
      IF(TORF)THEN
         READ(1,'(A)') FILENAME
         CALL READ_XMASK(XMASK,NCOL,NROW,FILENAME,
     $        IROW,ICOL)
      ELSE
         READ(1,*) FDUM
         CALL INIT_ARRAY(XMASK,NCOL,NROW,FDUM)
      ENDIF
c     read fraction file
      READ(1,*)
      READ(1,*)TORF
      IF(TORF)THEN
         READ(1,'(A)') FILENAME
         CALL READ_FRACTION(FRACTION,NCOL,NROW,FILENAME,
     $        IROW,ICOL)
      ELSE
         READ(1,*) FDUM
         CALL INIT_ARRAY(FRACTION,NCOL,NROW,FDUM)
      ENDIF
c     station file
      READ(1,'(/A)')FILENAME
      OPEN(10,FILE=FILENAME)
c     read input path and precision of VIC filenames
      READ(1,'(/A)')INPATH
      READ(1,*)DPREC
c     output pathname
      READ(1,'(/A)')OUTPATH
c     number of days to process
      READ(1,*)
c     start and end year/month from VIC simulation
      READ(1,*) START_YEAR, START_MO, STOP_YEAR, STOP_MO
c     calculate number of days & months in simulation
      M=START_MO
      Y=START_YEAR
      NMONTHS = 0
      NDAY=0
      DO J=START_MO,12*(STOP_YEAR-START_YEAR)+STOP_MO
        IF(M.EQ.2) THEN
           LP=isaleap(Y)
        ELSE
           LP=0
        ENDIF
        NDAY = NDAY+DAYS_IN_MONTH(M)+LP
        NMONTHS = NMONTHS + 1
        MO(NMONTHS) = M
        YR(NMONTHS) = Y
        M = M + 1
        IF (M .GT. 12) THEN
            M = 1
            Y  = Y + 1
        ENDIF
      END DO
      IF(NDAY.GT.DAYS) THEN
         PRINT*, 'IN ROUT.F RESET DAYS TO ', NDAY
         STOP
      ENDIF
      PRINT*,'NDAY = ',NDAY, ' NMONTHS = ',NMONTHS
c     start and end year/month for writing output
      READ(1,*) FIRST_YEAR, FIRST_MO, LAST_YEAR, LAST_MO
c     uh file
      READ(1,'(/A)')FILENAME

C***********************************************************

      CALL MAKE_UHM(UHM,VELO,DIFF,XMASK,NCOL,NROW,LE,DT,
     $        IROW,ICOL)

C     Loop over required stations

 100  CONTINUE
      READ(10,*,END=110)
     &     NR, NAME, PI, PJ, AREA
      READ(10,'(A80)',END=110) UH_STRING   !new, AW:  uh_string
      IF (NR .EQ. 1) THEN
         WRITE(*,'(I2,2X,A,I4,I4,G12.6)')
     &        NR, NAME, PI, PJ
         PRINT*, 'Routing station: ', NAME
c     note, the arrays are flipped left to right
         PI=ICOL+1-PI
         NAME5 = NAME

      print*, 'searching catchment...'
         CALL SEARCH_CATCHMENT
     &        (PI,PJ,DIREC,NCOL,NROW,
     &        NO_OF_BOX,CATCHIJ,PMAX,IROW,ICOL)

      print*, 'reading grid_UH...'
         CALL READ_GRID_UH
     &        (UH_BOX, KE, PMAX, NO_OF_BOX, CATCHIJ,FILENAME)

      print*, 'making grid UH...'
         CALL MAKE_GRID_UH
     &        (DIREC, NO_OF_BOX, UH_DAY, TMAX, PI, PJ, LE, UH_DAILY, KE,
     &        CATCHIJ,UHM, FR, PMAX, NCOL, NROW, UH_BOX, UH_S,
     &        UH_STRING,NAME5)        !new, AW:  added uh_string

      print*, 'making convolution...'
         CALL MAKE_CONVOLUTION
     &        (NCOL, NROW, NO_OF_BOX, PMAX, DAYS,
     &        CATCHIJ, BASE, RUNO, FLOW, KE, UH_DAY, UH_S, FRACTION,
     &        FACTOR_SUM,XC,YC,SIZE,DPREC,INPATH,ICOL,NDAY,
     &        IDAY,IMONTH,IYEAR, MO, YR, NYR)

      print*, 'writing data...'
         CALL WRITE_DATA
     &        (FLOW, NDAY, NAME5, FACTOR_SUM, OUTPATH,IDAY,IMONTH,IYEAR)

         CALL WRITE_MONTH
     &     (DAYS_IN_MONTH,START_YEAR, STOP_YEAR, FIRST_YEAR,
     &     LAST_YEAR, START_MO, STOP_MO, FIRST_MO,
     &     LAST_MO,
     &     NAME5, DAYS, FLOW, FACTOR_SUM, MONTHLY, MONTHLY_mm,
     &     YEARLY,YEARLY_mm,OUTPATH,NDAY,IMONTH,IYEAR,MO,YR,NMONTHS,NYR)


      END IF
      GOTO 100
 110  CONTINUE

      STOP
 9001 WRITE(*,*) 'CANNOT OPEN: ', FILE_INPUT
      END
c     ***********************************************
c     FUNCTION  ISALEAP

      integer function isaleap( iyr )

c     return 1 if a leap yr else 0

      if( (mod(iyr,4) .eq. 0 .and. mod(iyr,100) .ne.0)
     $                       .or. mod(iyr,400) .eq. 0) then
         isaleap = 1
      else
         isaleap = 0
      endif

      end
